sys/net/application_layer/telnet/telnet_server.c

```diff
@@ -0,0 +1,257 @@
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @brief       Telnet server implementation
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include <string.h>
+#include <fcntl.h>
+#include "net/sock/tcp.h"
+#include "net/telnet.h"
+#include "pipe.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#ifndef CONFIG_TELNET_TCP_QUEUE_SIZE
+#define CONFIG_TELNET_TCP_QUEUE_SIZE (1)
+#endif
+
+static char _stdin_pipe_buf[16];
+static ringbuffer_t _stdin_ringbuffer;
+static pipe_t _stdin_pipe;
+
+static sock_tcp_queue_t sock_queue;
+static sock_tcp_t socks[CONFIG_TELNET_TCP_QUEUE_SIZE];
+static sock_tcp_t *client;
+
+static char telnet_stack[THREAD_STACKSIZE_DEFAULT];
+
+#define SOCK_TCP_TIMEOUT_MS 50
+
+enum {
+    TELNET_CMD_EOF  = 236,
+    TELNET_CMD_SE   = 240,
+    TELNET_CMD_NOP,
+    TELNET_CMD_DATA_MARK,
+    TELNET_CMD_BRK,
+    TELNET_CMD_IP,
+    TELNET_CMD_AO,
+    TELNET_CMD_AYT,
+    TELNET_CMD_EC,
+    TELNET_CMD_EL,
+    TELNET_CMD_GA,
+    TELNET_CMD_SB   = 250,
+    TELNET_CMD_WILL,
+    TELNET_CMD_WONT,
+    TELNET_CMD_DO,
+    TELNET_CMD_DONT,
+    TELNET_CMD_IAC
+};
+
+enum {
+    TELNET_OPT_BINARY   = 0,
+    TELNET_OPT_ECHO     = 1,
+    TELNET_OPT_SUP_GO_AHEAD = 3,
+    TELNET_OPT_STATUS   = 5,
+};
+
+static bool connected;
+static mutex_t connected_mutex = MUTEX_INIT_LOCKED;
+static mutex_t sock_mutex;
+
+static void _acquire(void)
+{
+    mutex_lock(&sock_mutex);
+}
+
+static void _release(void)
+{
+    mutex_unlock(&sock_mutex);
+}
+
+static void _connected(void)
+{
+    connected = true;
+    mutex_unlock(&connected_mutex);
+}
+
+static void _disconnect(void)
+{
+    mutex_trylock(&connected_mutex);
+    connected = false;
+}
+
+static int _write_buffer(const void* buffer, size_t len)
+{
+    int res = 0;
+    const uint8_t *buf = buffer;
+    _acquire();
+
+    while (len) {
+        res = sock_tcp_write(client, buf, len);
+        if (res < 0) {
+            break;
+        }
+        len -= res;
+        buf += res;
+    }
+    _release();
+
+    return res < 0 ? res : 0;
+}
+
+static uint8_t _will(uint8_t option)
+{
+    switch (option) {
+    /* agree to suppress go-ahead packets */
+    /* see RFC 858 */
+    case TELNET_OPT_SUP_GO_AHEAD: return TELNET_CMD_DO;
+    }
+
+    return TELNET_CMD_WONT;
+}
+
+static void _process_cmd(uint8_t cmd, uint8_t option)
+{
+    DEBUG("cmd: %u, option: %u\n", cmd, option);
+    switch (cmd) {
+    case TELNET_CMD_WILL:
+        {
+        uint8_t reply[] = {TELNET_CMD_IAC, _will(option), option};
+        _write_buffer(reply, sizeof(reply));
+        break;
+        }
+    }
+}
+
+static void *telnet_thread(void *arg)
+{
+    (void)arg;
+
+    static uint8_t rx_buf[64];
+
+    while (1) {
+        ssize_t res = sock_tcp_accept(&sock_queue, &client, SOCK_NO_TIMEOUT);
+        if (res < 0) {
+            DEBUG("accept error: %s\n", strerror(-res));
+            continue;
+        }
+
+        DEBUG("connected\n");
+        _connected();
```

Comment:
**Iff** we go ahead with this, there should also be a big fat warning on connect towards the client, MOTD style.